### ISSUE TITLE
[Question] Long running processes with PAWS clients

### ISSUE THREAD

**Issue by mswerlick** [paws-r/paws#300]
I'm using PAWS to poll an SQS queue and post messages to SNS, a process which is long running. I'm using temporary credentials including a session token (valid for 12h) stored in a `.aws/credentials` file which is periodically over written with new credentials. After 12 hours,  I am getting `Error: ExpiredToken (HTTP 403). The security token included in the request is expired` (The process for re-generating these creds runs every 11 hours).

Is this expected behavior ( that being, methods from a client not working once the original credentials have expired)? Is there a way to force clients to refresh their credentials based on current values in either a credentials file or w/in env variables? Or is the best solution here to create new instances of the clients?

Thanks

**Comment by davidkretch** [comment]
Hmm, that is not expected. Sorry about that. It should look up the credentials every time the method runs, rather than when the client instances are created, but that is not happening here for some reason. Can you paste in an example (not necessarily working) of how the code is calling SQS?

**Comment by mswerlick** [comment]
Thanks for the quick reply. See below for a dumbed down version of what I'm trying to do. Basically I'm reading messages off of an SQS queue, passing the messages to application code, and then publishing the output to an SNS topic. I don't have credentials in environment variables and my credentials file looks like:

```
[default]
aws_access_key_id = XXXXXXXXXXX
aws_secret_access_key = XXXXXXXXXXXX
aws_session_token = XXXVERYLONGSTRINGXXXX
```

There are other moving parts with what I'm running, so if the below looks like it should be working then just knowing that PAWS consumes credentials when methods are called (not when client instances are created) is helpful in ruling out a potential issue. 

``` 
sqs <- paws::sqs()
sns <- paws::sns()

while(TRUE){
  # Poll the queue for up to 20 seconds
  # As soon as messages(s) are available they will be picked up
  # If no messages are found, NULL returned. Polling will start again

  messages <- sqs$receive_message(QueueUrl='myqueue-url', WaitTimeSeconds = 20)

  # If messages are found, run application code
  if(length(messages$Messages) > 0){
    for(req in messages$Messages){
      tryCatch(expr = {
        loginfo('Processing %s', req$Body)

        # Application code goes here
        response <- dummy_function(req)

        # If successful in rendering a message, post message to SNS
        loginfo('Message generated, posting message back to %s', integration$topic)
        sns$publish(TopicArn = 'arn:sometopic-etc',
                     Message = {'a message in json'}) 
        sqs$delete_message(QueueUrl = 'myqueue-url' , ReceiptHandle = req$RecieptHandle)
      },
      error = function(e){
        # If message handling (application code) or posting to SNS fails, log error
        # Return message to queue to be processed again
        message(e)

        loginfo('Error processing message: %s .Returning message to queue', req$ReceiptHandle)
        sqs$change_message_visibility(QueueUrl = 'myqueue-url,
                                      ReceiptHandle = req$ReceiptHandle,
                                      VisibilityTimeout = 0)
      })
    }
  }
}
```
**PAWS Info**
paws -  0.1.8
paws.common - 0.3.1

**System info**

```
platform       x86_64-pc-linux-gnu         
arch           x86_64                      
os             linux-gnu                   
system         x86_64, linux-gnu           
status                                     
major          3                           
minor          6.1                         
year           2019                        
month        07                          
day            05                          
svn rev        76782                       
language       R                           
version.string R version 3.6.1 (2019-07-05)
nickname       Action of the Toes  
```

**Comment by davidkretch** [comment]
Ok, sorry about that. Thank you for all the info. I'll take a look, likely towards the end of the week.

**Comment by mswerlick** [comment]
@davidkretch  I did some further debugging and this looks like its on my end. I had overlooked a part of my application code which setting environment variables for credentials (but only once). So once the initial token I generated had expired and been replaced in `.aws/credentials` my SQS client was still using the expired creds in the environment.

Apologies on this, and thanks for helping me narrow down where my problem was
